### PR TITLE
envsh: add option to watch workspace changes

### PR DESCRIPTION
**Pull request by g-arjones**
Pending issues:

 - [x] fix `#update_workspace`
 - [x] add tests

@doudou Thoughts? How do I fix `#update_workspace`?

### CODE REVIEW

**Comment by doudou** [comment]
First, given how little similarity there is between envsh and `envsh --watch`, I'd go for creating a new subcommand (possibly "advanced" so that it doesn't show up in the --help).

I'll have a look about `update_workspace`. The immediate problem I see is with the way autobuild uses Rake (which I should have fixed a long time ago) and with the global autobuild package registry (which should go away). Is there anything else ?

The quick & dirty way to avoid these issues would be to have autoproj exec itself when anything changes...

**Comment by g-arjones** [comment]
> I'd go for creating a new subcommand

Sounds good...

> with the global autobuild package registry

That's what's breaking the method...

> Is there anything else ?

Not that I know of.

> The quick & dirty way to avoid these issues would be to have autoproj exec itself when anything changes...

We still need an updated instsance of `Autoproj::Workspace` so we can update the watchers in case new packages/package sets get added..

**Comment by g-arjones** [comment]
> One thing, though, you would need to set up the watchers before the initial update, or there's a window where you would miss an update.

With `exec` there is no way to prevent that. Events triggered while the process is being replaced and before the watchers are setup _will_ be missed, regardless of how early the setting up happens. That said, the earliest, the better.

Will update.

**Comment by doudou** [comment]
> With exec there is no way to prevent that. Events triggered while the process is being replaced and before the watchers are setup will be missed, regardless of how early the setting up happens. That said, the earliest, the better.

Since you do an initial update in `watch`, this is not an issue:

$ autoproj watch
1. Setup watch
2. Initial update
3. Wait
4. Got event -> Exec
5. Setup watch
6. Initial update <- this update handles the event that triggered the exec *and* any event that might have happened between 4 and 5

**Comment by g-arjones** [comment]
> this update handles the event that triggered the exec and any event that might have happened between 4 and 5

Good point! 👍

File: lib/autoproj/cli/envsh.rb [hunk 1]
```diff
@@ -1,19 +1,147 @@
...
+                _, options = super(unused, options)
+                @watchers = []
+                @watch = options[:watch]
+                @options = options
                 options
             end
 
-            def run(**options)
+            def update_workspace(restart = false)
+                exec('autoproj', 'envsh', '--watch') if restart
```
Comment (doudou) [inline lib/autoproj/cli/envsh.rb]: This method is now doing two completely unrelated things based on the restart flag. Separate the methods into `update_workspace` and `restart` (and *definitely* split envsh from watch).

File: lib/autoproj/cli/envsh.rb [hunk 2]
```diff
@@ -1,19 +1,147 @@
...
+
+            def create_file_watcher(file)
+                watchers << notifier.watch(file, :modify) do |e|
+                    delete_watcher(e.watcher) if e.flags.include? :ignored
+                    callback
+                end
+            end
+
+            def create_dir_watcher(dir, *files, recursive: false)
+                opt_args = %i[move create delete]
```
Comment (doudou) [inline lib/autoproj/cli/envsh.rb]: I would prefer having the static arguments directly in the call for clarity:

`watchers << notify.watch(dir, :move, :create, :delete, *opt_args) ...`

**Comment by g-arjones** [comment]
I have to load the workspace before setting up the watchers otherwise I won't know what to watch :(

**Comment by doudou** [comment]
> I have to load the workspace before setting up the watchers otherwise I won't know what to watch :(

Ah well ... Let's assume nobody will manage to save just at the wrong time.

For what it's worth, I noticed something in Listen that might bite us here (I'm mentioning it so that we keep it in mind): they have a fair bit of logic dedicated to transforming an editor's atomic save (which is basically a write -> rename) into a single modify event. I'm guessing that this logic should handle it fine (it would trigger on rename), but I'm not entirely sure right now.

**Comment by g-arjones** [comment]
> I noticed something in Listen

Not using Listen anymore..

We have a dir watcher for every file watcher so I don't see how that could be a problem..

**Comment by doudou** [comment]
> Not using Listen anymore..

Which is why I'm mentioning it. I wouldn't mention it otherwise, as the complexity (fs-event-wise) of a simple file save is not hidden from us. A simple file save can lead to many events, and I want to write it down so that this is documented.

Simple example - which would not be so far-fetched if we weren't exec'ing - is that we would try to load the new workspace at a point where the file has been moved but not yet replaced, which would be invalid. Wouldn't happen often, but would happen.

A.k.a. there's always a cost to be close to the metal. You don't benefit from years of knowledge from people that are doing *exactly* what we're tying to do. Listen's main use-case is guard, which is basically a "run something when something changes" environment.

**Comment by doudou** [comment]
Looks good .. only missing tests now

**Comment by g-arjones** [comment]
Great.. I got caught by some "urgent" (?) stuff and could not finish them but they are super easy to write, autoproj's test suite already has all helpers I need I just did not have time to.. Will probably have to wait until after carnival now :P

**Comment by g-arjones** [comment]
Updated.

**Comment by doudou** [comment]
*Really* cool. Thanks a lot. I'd like to test it out here before I merge, but it looks good.

**Comment by g-arjones** [comment]
Sounds good. Thanks..

**Comment by doudou** [comment]
Houston, we have a problem ...

A lot of autoproj commands (namely, autoproj exec) are saving the config file which triggers watch ... and would probably trigger envsh which would save the config file :P. I'm looking into that.

**Comment by doudou** [comment]
.... and I already got concurrent file modification on `config.yml` ...

**Comment by g-arjones** [comment]
Damn.. Did not expect that. Thanks for having a look

**Comment by g-arjones** [comment]
> .... and I already got concurrent file modification on config.yml ...

That is something I would expect but seemed unlikely. How did you do that? We probably need a file lock

**Comment by doudou** [comment]
> That is something I would expect but seemed unlikely. How did you do that? We probably need a file lock

Since all commands write config.yml, it's a lot easier than it sounds. I had `autoproj watch` and Ruby linters running under `autoproj exec`

I'm working on all of this (and adding tests for Configuration ... this is really my cross to bear). Could you please try to create branches and pull requests directly on the main repositories ? It would make things a lot easier as I could push directly to the PR instead of having to create a new one.

I'm moving to use atomic writes instead of file locks for the time being

**Comment by g-arjones** [comment]
> Could you please try to create branches and pull requests directly on the main repositories 

Don't have write permission :(

**Comment by doudou** [comment]
> Don't have write permission :(

Then ask for it... I thought you did.

**Comment by doudou** [comment]
The "rock core" team somehow did not have write permissions on autoproj ... Just fixed it.

**Comment by g-arjones** [comment]
Would you prefer to close this and open another one?

**Comment by doudou** [comment]
> Would you prefer to close this and open another one?

Would be great. Just refer to this thread in the PR message.

**Pull request closed without merging**